File: concepts/06_text_binary_and_serialization/05_serialization_clone_and_transfer/cpp/test_02_numeric_and_trust_boundaries.cpp
// 数值模型、类型恢复与信任边界。
// 共同问题：通用数据格式是否保留语言数值类型；非标准数值如何处理；
// 自定义反序列化如何限制可构造类型和字段。
//
// polyglot-family: text_binary_and_serialization
// polyglot-concept: serialization_clone_and_transfer
// polyglot-related: languages/cpp/standard_library/13_strings_and_text/
// polyglot-related+: test_122_charconv_numeric_conversion_and_cxx20_formatting.cpp
// polyglot-related: languages/cpp/language/test_017_object_representation_alignment_and_bit_cast.cpp

#include <gtest/gtest.h>

#include <array>
#include <charconv>
#include <cstdint>
#include <span>
#include <stdexcept>
#include <string>

namespace {

std::int64_t decode_integer(const std::string& text) {
  std::int64_t value = 0;
  const auto result = std::from_chars(text.data(), text.data() + text.size(), value);
  if (result.ec != std::errc{} || result.ptr != text.data() + text.size()) {
    throw std::invalid_argument{"invalid integer"};
  }
  return value;
}

struct Packet {
  std::int8_t value;
};

Packet decode_packet(std::span<const std::uint8_t> bytes) {
  constexpr std::uint8_t supported_version = 1;
  constexpr std::uint8_t point_tag = 1;
  if (bytes.size() != 3 ||
      bytes[0] != supported_version ||
      bytes[1] != point_tag) {
    throw std::invalid_argument{"unsupported packet"};
  }
  return Packet{static_cast<std::int8_t>(bytes[2])};
}

TEST(SerializationBoundaryConcept, ExplicitIntegerParserPreservesInt64Value) {
  const std::string text = "9007199254740993";

  EXPECT_EQ(decode_integer(text), INT64_C(9007199254740993));
  EXPECT_THROW(decode_integer("9007199254740993tail"), std::invalid_argument);

  // C++20 没有标准 JSON 类型；from_chars 按目标 int64_t 的范围精确解析，不会经过
  // JavaScript Number。协议仍须声明目标范围和溢出策略。
}

TEST(SerializationBoundaryConcept, DecoderWhitelistsLengthVersionAndTypeTag) {
  const std::array<std::uint8_t, 3> valid{1, 1, 0xF9};
  const std::array<std::uint8_t, 2> short_input{1, 1};
  const std::array<std::uint8_t, 4> trailing{1, 1, 7, 0};
  const std::array<std::uint8_t, 3> future_version{2, 1, 7};
  const std::array<std::uint8_t, 3> unknown_type{1, 9, 7};

  EXPECT_EQ(decode_packet(valid).value, -7);
  EXPECT_THROW(decode_packet(short_input), std::invalid_argument);
  EXPECT_THROW(decode_packet(trailing), std::invalid_argument);
  EXPECT_THROW(decode_packet(future_version), std::invalid_argument);
  EXPECT_THROW(decode_packet(unknown_type), std::invalid_argument);

  // 先验证完整边界再解释字段；不要按不可信标签构造任意类型，也不要直接持久化对象布局。
}

}  // namespace

File: concepts/06_text_binary_and_serialization/05_serialization_clone_and_transfer/rust/test_02_numeric_and_trust_boundaries.rs
// polyglot-family: text_binary_and_serialization
// polyglot-concept: serialization_clone_and_transfer
// polyglot-related: languages/rust/tests/standard_library/
// polyglot-related+: 14_formatting_parsing_binary_and_serialization/test_110_size_prefix_and_trailing_data_validation.rs
//
// 共同问题：大整数和特殊浮点怎样编码；未知字段、长度、cycle 与不可信输入在哪里拒绝。
// 对照观察：没有内建 JSON number model；manual codec 必须在 allocation 前检查 prefix 并定义精度与 schema。

fn decode(bytes: &[u8], maximum: usize) -> Result<&[u8], &'static str> {
    let prefix: [u8; 2] = bytes.get(..2).ok_or("missing length")?.try_into().unwrap();
    let length = u16::from_be_bytes(prefix) as usize;
    if length > maximum {
        return Err("too large");
    }
    let payload = bytes.get(2..2 + length).ok_or("truncated")?;
    if bytes.len() != 2 + length {
        return Err("trailing bytes");
    }
    Ok(payload)
}

#[test]
fn comparison() {
    assert_eq!(decode(&[0, 2, b'o', b'k'], 8), Ok(&b"ok"[..]));
    assert_eq!(decode(&[0, 9], 8), Err("too large"));
    assert_eq!(decode(&[0, 2, b'o'], 8), Err("truncated"));
    assert_eq!(
        9_007_199_254_740_993_u64 as f64 as u64,
        9_007_199_254_740_992
    );
}
